x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -874,8 +904,19 @@ private static List<Attribute> potentialCandidatesIfNoMatchesFound(
         return matches;
     }
 
-    private static Attribute handleSpecialFields(UnresolvedAttribute u, Attribute named) {
-        return named.withLocation(u.source());
+    private static Attribute tryResolveUnionTypeFields(UnresolvedAttribute toResolve, Expression parent, Attribute resolveTo) {
+        if (resolveTo instanceof FieldAttribute fa) {
+            // incompatible mappings
+            var field = fa.field();
+            // Multi-typed fields are allowed if and only if we convert them to a single type immediately.
+            if (field instanceof InvalidMappedField imf && (parent instanceof AbstractConvertFunction == false)) {
+                resolveTo = toResolve.withUnresolvedMessage(
+                    "Cannot use field [" + fa.name() + "] due to ambiguities being " + imf.errorMessage()
+                );
+            }
+        }
+
+        return resolveTo.withLocation(toResolve.source());
```

Comment:
`handleSpecialFields` used to look for FieldAttributes mapped with multiple types before https://github.com/elastic/elasticsearch/pull/107545. This brings back this logic, but takes into account the attribute's parent because multi-typed FieldAttributes are allowed if they're immediately cast.